src/nvim/eval/vars.c

```diff
@@ -766,84 +766,84 @@ static char *ex_let_option(char *arg, typval_T *const tv, const bool is_const,
-    varnumber_T n = 0;
-    getoption_T opt_type;
-    long numval;
-    char *stringval = NULL;
-    const char *s = NULL;
-    bool failed = false;
-    uint32_t opt_p_flags;
-    char *tofree = NULL;
-
-    const char c1 = *p;
-    *p = NUL;
+    return NULL;
+  }
 
-    opt_type = get_option_value(arg, &numval, &stringval, &opt_p_flags, scope);
-    if (opt_type == gov_bool
-        || opt_type == gov_number
-        || opt_type == gov_hidden_bool
-        || opt_type == gov_hidden_number) {
-      // number, possibly hidden
-      n = (long)tv_get_number(tv);
-    }
-
-    if ((opt_p_flags & P_FUNC) && tv_is_func(*tv)) {
-      // If the option can be set to a function reference or a lambda
-      // and the passed value is a function reference, then convert it to
-      // the name (string) of the function reference.
-      s = tofree = encode_tv2string(tv, NULL);
-    } else if (tv->v_type != VAR_BOOL && tv->v_type != VAR_SPECIAL) {
-      // Avoid setting a string option to the text "v:false" or similar.
-      s = tv_get_string_chk(tv);
-    }
-
-    if (op != NULL && *op != '=') {
-      if (((opt_type == gov_bool || opt_type == gov_number) && *op == '.')
-          || (opt_type == gov_string && *op != '.')) {
-        semsg(_(e_letwrong), op);
-        failed = true;  // don't set the value
-      } else {
-        // number or bool
-        if (opt_type == gov_number || opt_type == gov_bool) {
-          switch (*op) {
-          case '+':
-            n = numval + n; break;
-          case '-':
-            n = numval - n; break;
-          case '*':
-            n = numval * n; break;
-          case '/':
-            n = num_divide(numval, n); break;
-          case '%':
-            n = num_modulus(numval, n); break;
-          }
-          s = NULL;
-        } else if (opt_type == gov_string && stringval != NULL && s != NULL) {
-          // string
-          char *const oldstringval = stringval;
-          stringval = concat_str(stringval, s);
-          xfree(oldstringval);
-          s = stringval;
+  bool hidden;
+  bool error;
+  const char c1 = *p;
+  *p = NUL;
+
+  OptVal curval = get_option_value(arg, NULL, scope, &hidden);
+  OptVal newval = tv_to_optval(tv, arg, scope, &error);
+
+  // Ignore errors for num types
+  if (newval.type != kOptValTypeNumber && newval.type != kOptValTypeBoolean && error) {
+    goto end;
+  }
+
+  // Don't assume current and new values are of the same type in order to future-proof the code for
+  // when an option can have multiple types.
+  const bool is_num = ((curval.type == kOptValTypeNumber || curval.type == kOptValTypeBoolean)
+                       && (newval.type == kOptValTypeNumber || newval.type == kOptValTypeBoolean));
+  const bool is_string = curval.type == kOptValTypeString && newval.type == kOptValTypeString;
+
+  if (op != NULL && *op != '=') {
+    if (!hidden && ((is_num && *op == '.') || (is_string && *op != '.'))) {
+      semsg(_(e_letwrong), op);
+      goto end;
```

Comment:
If there will be options with multiple types in future, the type of the new value should have to depend on both the type of the old value and the operator used, so deciding the type of the new value here before checking for the operator doesn't look right.

Reply:
I think it better to pass the operator type to `tv_to_optval()` and check for it there. This may also allow changing back the error number in `test_vimscript.vim`.

Reply:
> I think it's better to pass the operator `op` to `tv_to_optval()` and check for it there. This may also allow changing back the error number in `test_vimscript.vim`.

well I think we add a `keep_type` boolean flag instead which'd make it try to keep the old value's type for the new value. It's more general since `tv_to_optval` is used in `eval_option()` as well

Reply:
Why keep type when an operator is specified? What if there will type changes like the change between number and bool?

Reply:
It's general in the sense that `set_option_from_tv` doesn't use an operator and thus will have no operator, and `ex_let_option` can just handle the operator itself to set the value of the boolean flag that it passes to `tv_to_optval`.

> What if there will type changes like the change between number and bool

Wdym? That'll just be handled normally

Reply:
> It's general in the sense that `set_option_from_tv` doesn't use an operator and thus will have no operator, and `ex_let_option` can just handle the operator itself to set the value of the boolean flag that it passes to `tv_to_optval`.
> 
> > What if there will type changes like the change between number and bool
> 
> Wdym? That'll just be handled normally. If the current value is a bool, the new value will be too. Same for number option value.

Actually, no, numeric types won't be changed. So if new value is number and old value is bool, new value will remain a number. Num and bool types are handled together in `tv_to_optval` anyway so that shouldn't be an exceptional case. Besides, the `typval` stuff is stuff we can refactor again later down the line once we have a better sense of how we want to design options in general.

Reply:
Actually I think the following makes the more sense:

Add a `want_type` argument to force `tv_to_optval()` to get that type. If `want_type` is provided `tv_to_optval()` also doesn't need to call `get_option_value()` again.

Reply:
> Actually I think the following makes more sense:
> 
> Add a `want_type` argument to force `tv_to_optval()` to get that type. If `want_type` is provided `tv_to_optval()` also doesn't need to call `get_option_value()` again.

I really like this as well. I think it's better than my idea.

Reply:
It seems that using a wrong type for an option gives multiple error messages in many cases in Vim. Maybe it's better to fix that in Vim first.

Reply:
> It seems that using a wrong type for an option gives multiple error messages in many cases in Vim. Maybe it's better to fix that in Vim first.

Could you give me an example

Reply:
`:let &number = []`
```
E745: Using a List as a Number
E730: Using a List as a String
```

Reply:
I'll send a PR to Vim after deciding what to do.